project/EngineNativeLibraryExtractor.scala

```diff
@@ -0,0 +1,176 @@
+import sbt.*
+import sbt.util.CacheStoreFactory
+
+object EngineNativeLibraryExtractor {
+  private val JLINE_NATIVE = "jline-native"
+
+  /** Extracts native libraries from engine jars (all transitive dependencies of
+    * engine).
+    *
+    * For every JAR file for which native libraries should be extracted:
+    * <ul>
+    *   <li>Extracts the native libraries into the `componentDir` directory.</li>
+    *   <li>Copies the rest of the JAR archive into a new JAR file with the
+    *   `-thin.jar` suffix, which contains only the non-native parts of the JAR.</li>
+    *   <li>Deletes the original (fat) JAR file from the `componentDir` directory.</li>
+    * </ul>
+    * @param componentDir Component directory where all the dependencies of engine are stored.
+    * @param updateReport Update report for the "engine-runner" project.
+    *            Get it with `(engine-runner/update).value`.
+    * @param scalaBinaryVersion
+    * @param cacheFactory
+    * @param previousRun
+    * @return
+    */
+  def extractNativeLibraries(
+    componentDir: File,
+    logger: Logger,
+    updateReport: UpdateReport,
+    scalaBinaryVersion: String,
+    cacheFactory: CacheStoreFactory,
+    previousRun: Option[AnalysisOfExtractedNativeLibs]
+  ): AnalysisOfExtractedNativeLibs = {
+    val jlineNativeJars = JPMSUtils.filterModulesFromUpdate(
+      updateReport,
+      Dependencies.jlineNative,
+      logger,
+      "engine-runner",
+      scalaBinaryVersion,
+      shouldContainAll = true
+    )
+    require(
+      jlineNativeJars.size == 1,
+      s"Expected exactly one jline-native jar, found: ${jlineNativeJars}."
+    )
+    val jlineNativeJar = jlineNativeJars.head
+    var analysis = previousRun
+      .getOrElse(AnalysisOfExtractedNativeLibs(Map.empty))
+    if (!componentDir.exists()) {
+      logger.error(
+        s"Component directory [$componentDir] does not exist yet. " +
+        "It must have been created with createEnginePackageNoIndex task."
+      )
+      throw new IllegalStateException("Component directory does not exist.")
+    }
+    logger.debug(
+      s"[EngineNativeLibraryExtractor] Extracting native libraries. " +
+      s"jlineNativeJar: $jlineNativeJar, analysis; $analysis"
+    )
+
+    getFromCache(jlineNativeJar, analysis) match {
+      case Some(cachedLibs) =>
+        logger.debug(
+          "jline-native libraries already extracted, skipping extraction."
+        )
+      case None =>
+        val outJarName = jlineNativeJar.getName.replace(
+          ".jar",
+          "-thin.jar"
+        )
+        val outJar = componentDir / outJarName
+        if (outJar.exists()) {
+          // Extraction is not cache, but the output jar exist. We must first delete it.
+          // This can happen if sbt was restarted and the outputJar was not deleted.
+          IO.delete(outJar)
```

Comment:
Similar trick is used, e.g., in [std-image/cleanPolyglotRoot](https://github.com/enso-org/enso/blob/774b1f9f9302b5144e9dccb59b327c2ac507f229/build.sbt#L4910) - it is possible that the old file exist on the disk, but it is not in the cache - for example if sbt was restarted. In that case, just ensure that the file is deleted.